gematria/datasets/annotating_importer.cc

```diff
@@ -0,0 +1,340 @@
+
+  // Find the relevant mapping.
+  quipper::PerfDataProto::MMapEvent mmap;
+  for (const auto &event : perf_data_proto.events()) {
+    // TODO(virajbshah): Not sure if this always works, i.e. does the main
+    // binary always correspond to the first MMapEvent. Implement BuildID or
+    // name based checking.
+    if (event.has_mmap_event() &&
+        event.mmap_event().prot() & 1 /* PROT_READ */ &&
+        event.mmap_event().prot() & 4 /* PROT_EXEC */) {
+      mmap = event.mmap_event();
+      break;
+    }
+  }
+  // TODO(virajbshah): Make sure the mapping was found. (Use num_mmap_events)
+
+  // If the profile has multiple event types, lookups are needed to find the
+  // event type corresponding to a sample. In the other case, this is neither
+  // required nor possible - since samples are not associated with IDs for
+  // lookup in the proto.
+  const bool has_multiple_sample_types = sample_types.size() > 1;
+
+  // Process sample events.
+  std::unordered_map<uint64_t, std::vector<int>> samples;
+  for (const auto &event : perf_data_proto.events()) {
+    // Filter out non-sample events.
+    if (!event.has_sample_event()) {
+      continue;
+    }
+
+    // Filter out sample events from outside the profiled binary.
+    uint64_t sample_ip = event.sample_event().ip();
+    if (!(mmap.start() <= sample_ip && sample_ip < mmap.start() + mmap.len())) {
+      continue;
+    }
+
+    if (!samples.count(sample_ip)) {
+      samples[sample_ip] = std::vector<int>(sample_types.size(), 0);
+    }
+    int event_idx = 0;
+    if (has_multiple_sample_types) {
+      event_idx =
+          event_code_to_idx[event_id_to_code[event.sample_event().id()]];
+    }
+    samples[sample_ip][event_idx] += 1;
+  }
+
+  return make_pair(sample_types, samples);
+}
+
+absl::StatusOr<
+    std::unordered_map<uint64_t, std::pair<uint64_t, std::vector<uint32_t>>>>
+AnnotatingImporter::GetLBRData() {
+  const quipper::PerfDataProto &perf_data_proto = perf_reader_.proto();
+
+  std::unordered_map<uint64_t, std::pair<uint64_t, std::vector<uint32_t>>>
+      lbr_data;
+  for (const auto &event : perf_data_proto.events()) {
+    if (!event.has_sample_event() ||
+        !event.sample_event().branch_stack_size()) {
+      continue;
+    }
+    const auto &branch_stack = event.sample_event().branch_stack();
+    for (int branch_idx = 0; branch_idx + 1 < branch_stack.size();
+         ++branch_idx) {
+      const auto &next_branch_entry = branch_stack.at(branch_idx);
+      const auto &branch_entry = branch_stack.at(branch_idx + 1);
+      if (lbr_data.count(branch_entry.to_ip())) {
+        auto &[run_end, cycles_values] = lbr_data[branch_entry.to_ip()];
+        if (next_branch_entry.from_ip() == run_end) {
+          cycles_values.push_back(next_branch_entry.cycles());
+        } else if (next_branch_entry.from_ip() < run_end) {
+          run_end = next_branch_entry.from_ip();
+          cycles_values = std::vector<uint32_t>{next_branch_entry.cycles()};
+        }
```

Comment:
As noted in the .h file: we could also preserve all the versions as separate entries. They will have some not taken conditional branches, but that makes them quite interesting for the training.